rpc/src/endpoint/evidence.rs

```diff
@@ -1,21 +1,87 @@
 //! `/broadcast_evidence`: broadcast an evidence.
 
-use serde::{Deserialize, Serialize};
+use serde::{ser::SerializeStruct, Deserialize, Serialize};
 use tendermint::{evidence::Evidence, Hash};
 
-use crate::{dialect::Dialect, request::RequestMessage, Method};
+use crate::{client::CompatMode, dialect::Dialect, request::RequestMessage, Method};
 
 /// `/broadcast_evidence`: broadcast an evidence.
-#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[derive(Clone, Debug, PartialEq, Eq)]
 pub struct Request {
     /// Evidence to broadcast
     pub evidence: Evidence,
+
+    pub compat_mode: CompatMode,
+}
+
+// This is a workaround for the fact that the serialization of the evidence
+// changed between Tendermint 0.34 and 0.37.
+impl Serialize for Request {
+    fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        match self.compat_mode {
+            CompatMode::V0_34 => {
+                use tendermint_proto::v0_34::types::Evidence as RawEvidence;
+                let mut state = serializer.serialize_struct("Request", 1)?;
+                state.serialize_field("evidence", &RawEvidence::from(self.evidence.clone()))?;
+                state.end()
+            },
+            CompatMode::V0_37 => {
+                use tendermint_proto::v0_37::types::Evidence as RawEvidence;
+                let mut state = serializer.serialize_struct("Request", 2)?;
+                state.serialize_field("evidence", &RawEvidence::from(self.evidence.clone()))?;
+                state.end()
+            },
+        }
+    }
+}
+
+// This is a terrible hack for the fact that the serialization of the evidence
+// changed between Tendermint 0.34 and 0.37.
+impl<'de> Deserialize<'de> for Request {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        use serde::de::Error;
+
+        #[derive(Deserialize)]
+        struct InnerV0_34 {
+            evidence: tendermint_proto::v0_34::types::Evidence,
+        }
+
+        #[derive(Deserialize)]
+        struct InnerV0_37 {
+            evidence: tendermint_proto::v0_37::types::Evidence,
+        }
```

Comment:
This is already close to what I'm talking about above, but it does not use generics and only works in `Deserialize` direction.